tests/PHPStan/Rules/Methods/data/bug-9542.php

```diff
@@ -0,0 +1,54 @@
+<?php declare(strict_types = 1);
+
+namespace Bug9542;
+
+use function PHPStan\Testing\assertType;
+
+interface TranslatableInterface
+{
+
+}
+
+class TranslatableMessage implements TranslatableInterface
+{
+	public function getMessage(): void
+	{
+		return;
+	}
+}
+
+class HelloWorld
+{
+	public function testInstanceOf(TranslatableInterface $translatable): void
+	{
+		if (! $translatable instanceof TranslatableMessage) {
+			assertType('Bug9542\TranslatableInterface~Bug9542\TranslatableMessage', $translatable);
+			return;
+		}
+
+		assertType('Bug9542\TranslatableMessage', $translatable);
+		$translatable->getMessage();
+	}
+
+	public function testClass(TranslatableInterface $translatable): void
+	{
+		if ($translatable::class !== TranslatableMessage::class) {
+			assertType('Bug9542\TranslatableInterface~Bug9542\TranslatableMessage', $translatable);
```

Comment:
There's a bad bug here. So you've eliminated that `$translatable` isn't `TranslatableMessage`. But it can still be a subclass of `TranslatableMessage`.

Most likely we should only specify the type for `$context->true()`.

Same thing is done for `get_class`: https://github.com/phpstan/phpstan-src/blob/2df1a5904a218038187e180861f757bb938a37b4/src/Analyser/TypeSpecifier.php#L194-L210

Reply:
could we be "smarter" if `TranslatableMessage` is a reference to a final class (or a interface)?

Reply:
We could, but first we need to get rid of the bug, then we can improve it :)

Reply:
adapted. better? I didn't even noticed this.. :)

Reply:
oh sorry, I was lazy and didn't run the full tests :/

Reply:
thx for the fix!